### ISSUE TITLE
how to get conditional embedding in inference stage.

### ISSUE THREAD

**Issue by dulvqingyun** [TencentYoutuResearch/PersonReID-YouReID#4]
Hello,  it seems during evaluation we only get the individual features of query set and gallery set,  i get map 0.836 on dataset market1501, and if I set use_rerank: true the map rise to 0.8945.
 From your paper section3.5 Model Inference, after get individual features of query set and gallery set, we need re calculate the distance  between query and gallery based the conditional embedding.  but from the code it seems that we discard the key-point Alignment Block and Conditional  Feature Embedding Block during inference, so could you tell me the reason?

**Comment by poorfriend** [comment]
> Hello, it seems during evaluation we only get the individual features of query set and gallery set, i get map 0.836 on dataset market1501, and if I set use_rerank: true the map rise to 0.8945.
> From your paper section3.5 Model Inference, after get individual features of query set and gallery set, we need re calculate the distance between query and gallery based the conditional embedding. but from the code it seems that we discard the key-point Alignment Block and Conditional Feature Embedding Block during inference, so could you tell me the reason?

1. 83.6?, maybe you use a wrong config. you can get a result above [this](https://github.com/TencentYoutuResearch/PersonReID-CACENET/issues/1#issuecomment-767414779) at least.
2. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set

**Comment by dulvqingyun** [comment]
if I set use_rerank: true I get map 0.8945,  it's close to 0.8963.

**Comment by dulvqingyun** [comment]
> 
> 
> > Hello, it seems during evaluation we only get the individual features of query set and gallery set, i get map 0.836 on dataset market1501, and if I set use_rerank: true the map rise to 0.8945.
> > From your paper section3.5 Model Inference, after get individual features of query set and gallery set, we need re calculate the distance between query and gallery based the conditional embedding. but from the code it seems that we discard the key-point Alignment Block and Conditional Feature Embedding Block during inference, so could you tell me the reason?
> 
>     1. 83.6?, maybe you use a wrong config. you can get a result above [this](https://github.com/TencentYoutuResearch/PersonReID-CACENET/issues/1#issuecomment-767414779) at least.
> 
>     2. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set

the paper get map 90.3 with re_rank or without re_rank?

**Comment by dulvqingyun** [comment]
> 
> 
> > Hello, it seems during evaluation we only get the individual features of query set and gallery set, i get map 0.836 on dataset market1501, and if I set use_rerank: true the map rise to 0.8945.
> > From your paper section3.5 Model Inference, after get individual features of query set and gallery set, we need re calculate the distance between query and gallery based the conditional embedding. but from the code it seems that we discard the key-point Alignment Block and Conditional Feature Embedding Block during inference, so could you tell me the reason?
> 
>     1. 83.6?, maybe you use a wrong config. you can get a result above [this](https://github.com/TencentYoutuResearch/PersonReID-CACENET/issues/1#issuecomment-767414779) at least.
> 
>     2. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set

could you tell me some details about  getting  conditional embedding feature in inference stage?  i got the query and gallery feature map respectively， but do not know how to get the conditional embedding feature. From the code  output of self._pair_graph_() is slice to global_feat_0 and global feat_1, so which one of them should be selected as conditional embedding feature?

**Comment by poorfriend** [comment]
> > > Hello, it seems during evaluation we only get the individual features of query set and gallery set, i get map 0.836 on dataset market1501, and if I set use_rerank: true the map rise to 0.8945.
> > > From your paper section3.5 Model Inference, after get individual features of query set and gallery set, we need re calculate the distance between query and gallery based the conditional embedding. but from the code it seems that we discard the key-point Alignment Block and Conditional Feature Embedding Block during inference, so could you tell me the reason?
> > 
> > 
> > ```
> > 1. 83.6?, maybe you use a wrong config. you can get a result above [this](https://github.com/TencentYoutuResearch/PersonReID-CACENET/issues/1#issuecomment-767414779) at least.
> > 
> > 2. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set
> > ```
> 
> the paper get map 90.3 with re_rank or without re_rank?
 without re_rank

**Comment by dulvqingyun** [comment]
@poorfriend , I can not replicate the result. I downloaded the pretrained resnet and dataset you provide and not change the  parameters in experiment/graph/cacenet.yaml。 I checked the log you provide and mine, they seems very close. the attached is my training log, could you help to see it?
[log.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6066057/log.txt)

**Comment by poorfriend** [comment]
> @poorfriend , I can not replicate the result. I downloaded the pretrained resnet and dataset you provide and not change the parameters in experiment/graph/cacenet.yaml。 I checked the log you provide and mine, they seems very close. the attached is my training log, could you help to see it?
> [log.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6066057/log.txt)

what is the map of this model.  Also 0.836?  I can't see any mistake from your log. what about train_mode in [eval_config](https://github.com/TencentYoutuResearch/PersonReID-CACENET/blob/a1cd4580b7d82d001b89a427da53442cbb27f92d/eval_config.yaml#L13),  it must be false.  I also have got a lower result when set train_mode true.

**Comment by dulvqingyun** [comment]
@poorfriend ,  the best map i get is 0.85， i set train mode to false. could you help to see my config file?  Thank you so much~~
[eval_config.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6073473/eval_config.txt)
[cacenet.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6073520/cacenet.txt)

**Comment by poorfriend** [comment]
> @poorfriend , the best map i get is 0.85， i set train mode to false. could you help to see my config file? Thank you so much~~
> [eval_config.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6073473/eval_config.txt)
> [cacenet.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6073520/cacenet.txt)

in cacenet.txt, non_local_block is 3, but in eval_config.txt, there is no non_local_block,  so it is set to default value 2.  May be it cause lower mAP

**Comment by dulvqingyun** [comment]
> > @poorfriend , the best map i get is 0.85， i set train mode to false. could you help to see my config file? Thank you so much~~
> > [eval_config.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6073473/eval_config.txt)
> > [cacenet.txt](https://github.com/TencentYoutuResearch/PersonReID-CACENET/files/6073520/cacenet.txt)
> 
> in cacenet.txt, non_local_block is 3, but in eval_config.txt, there is no non_local_block, so it is set to default value 2. May be it cause lower mAP

@poorfriend many thanks~~, after change non_local_block I get map 90% withou rerank.

**Comment by sophieyl** [comment]
> 2. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set

hi~
I want to know that is it all the results shown in your paper are testing without the inference of Conditional Feature?
how about the result of the  Occluded Duke? Can it achieve mAP to 50.8% only with extracting individual features?

@poorfriend

**Comment by sophieyl** [comment]
> > 1. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set
> 
> hi~
> I want to know that is it all the results shown in your paper are testing without the inference of Conditional Feature?
> how about the result of the Occluded Duke? Can it achieve mAP to 50.8% only with extracting individual features?
> 
> @poorfriend

I reproduce the result on Occluded Duke without changing any args, which mAP can reach to 53.97%

**Comment by poorfriend** [comment]
> > > 1. About Conditional Feature, we don't use Conditional Feature in this implement because our further experiments show that extracting the conditional embedding on little queries with lowest top-1 similarity and ranking rest of the queries with the CACENET's individual feature already achieves CACENET's current performance. In others words,, this observation shows that the conditional embedding training branch also improves the performance of CACENET's individual feature, and the conditional feature itself is only responsible for solving the hardest cases in the test set
> > 
> > 
> > hi~
> > I want to know that is it all the results shown in your paper are testing without the inference of Conditional Feature?
> > how about the result of the Occluded Duke? Can it achieve mAP to 50.8% only with extracting individual features?
> > @poorfriend
> 
> I reproduce the result on Occluded Duke without changing any args, which mAP can reach to 53.97%

tql
